File: src/core/golang/core/shell.go
package core

import (
	"time"
)

type ShellTypeId int32

const (
	BashShellId  ShellTypeId = 1
	PowerShellId             = 2
)

type ShellScript struct {
	Id          int64       `db:"id"`
	OrgId       int32       `db:"org_id"`
	TypeId      ShellTypeId `db:"type_id"`
	Name        string      `db:"name"`
	Description string      `db:"description"`
	BucketId    string      `db:"bucket_id"`
	StorageId   string      `db:"storage_id"`
}

type ShellScriptVersion struct {
	Id            int64     `db:"id"`
	ShellId       int64     `db:"shell_id"`
	OrgId         int32     `db:"org_id"`
	UploadTime    time.Time `db:"upload_time"`
	UploadUserId  int64     `db:"upload_user_id"`
	GcsGeneration int64     `db:"gcs_generation"`
}

type ShellScriptRun struct {
	Id              int64     `db:"id"`
	ScriptVersionId int64     `db:"script_version_id"`
	RunUserId       int64     `db:"run_user_id"`
	CreateTime      time.Time `db:"create_time"`
	RunTime         NullTime  `db:"run_time"`
	EndTime         NullTime  `db:"end_time"`
}

type ShellScriptRunPerServer struct {
	RunId        int64      `db:"run_id"`
	OrgId        int32      `db:"org_id"`
	ServerId     int64      `db:"server_id"`
	EncryptedLog NullString `db:"encrypted_log"`
	RunTime      NullTime   `db:"run_time"`
	EndTime      NullTime   `db:"end_time"`
	Success      bool       `db:"success"`
}

File: src/core/jsui/ts/shell.ts
export enum ShellTypes {
    Bash = 1,
    PowerShell = 2,
}

export let ShellTypeToCodeMirror = new Map<ShellTypes, string>()
ShellTypeToCodeMirror.set(ShellTypes.Bash, "text/x-sh")
ShellTypeToCodeMirror.set(ShellTypes.PowerShell, "application/x-powershell")

export interface ShellScript {
    Id : number
    OrgId : number
    TypeId : number
    Name : string
    Description : string
    BucketId : string
    StorageId: string
}

export interface ShellScriptVersion {
    Id: number
    ShellId: number
    OrgId: number
    UploadTime: Date
    UploadUserId : number
    GcsGeneration : number
}

export function cleanShellScriptVersionFromJson(v : ShellScriptVersion) {
    v.UploadTime = new Date(v.UploadTime)
}

export interface ShellScriptRun {
    Id: number
    ScriptVersionId: number
    RunUserId: number
    CreateTime: Date
    RunTime: Date | null
    EndTime: Date | null
}

export function cleanShellScriptRunFromJson(r : ShellScriptRun) {
    r.CreateTime = new Date(r.CreateTime)
    if (!!r.RunTime) {
        r.RunTime = new Date(r.RunTime)
    }

    if (!!r.EndTime) {
        r.EndTime = new Date(r.EndTime)
    }
}

export interface ShellScriptRunPerServer {
    RunId: number
    OrgId: number
    ServerId: number
    EncryptedLog: string | null
    RunTime: Date | null
    EndTime: Date | null
    Success: boolean
}

export function cleanShellScriptRunPerServerFromJson(r : ShellScriptRunPerServer) {
    if (!!r.RunTime) {
        r.RunTime = new Date(r.RunTime)
    }

    if (!!r.EndTime) {
        r.EndTime = new Date(r.EndTime)
    }
}
